_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2334 @@
+  infrastructure. To illustrate this point, consider a scenario where you are working on updating the cross account IAM
+  roles to access your environments, and you are at the point of reflecting your changes to prod. Would you want to
+  deploy that code to your live production environment off of an unreviewed branch? +
+  The only real test you can do for live infrastructure config is to do a dry run of your infrastructure code. Most Infrastructure as Code tools support a dry run of the code to check what it would do against your environment. For example, with Terraform, you could run `terraform plan` to sanity check the planned actions Terraform will take. This is especially useful for sanity checking a fresh clone of the code. The trunk should be a true reflection of the live environment, so you should expect there to be no changes to make on a fresh clone of trunk.
+
+
+[[make_code_changes]]
+==== Make code changes
+
+Now that you have a working local copy, you can start to make changes to the code. This process is done iteratively
+while checking for validity of the changes along the way with manual or automated testing. It is important to invest
+some time and effort in making the feedback cycle short, as it directly translates to your development speed. The faster
+you can iterate, the more tests you can run, and the better your code will be.
+
+How you make changes to the code will be largely the same for the three flavors of code we covered, although how you
+test your changes and the test cycles will be different. Typically, testing application code can be done in seconds
+(because everything is local), and testing live infrastructure config can be done in minutes (because you are only doing
+a dry run). However, testing infrastructure modules can take a long time since you need to deploy infrastructure (on the
+order of 10s of minutes). For ideas on how to improve the test cycles for infrastructure modules, take a look at
+https://terratest.gruntwork.io/docs/testing-best-practices/iterating-locally-using-test-stages/[Iterating locally using
+test stages] in the Terratest documentation.
+
+
+[[submit_changes_for_review]]
+==== Submit changes for review
+
+Once the code implementation is done and the testing passes, the next step is to submit it for review. You want to focus
+your review process on things that are hard to check through automated testing, such as checking security flaws,
+reviewing general code design, enforcing style guides, or identifying potential performance issues on larger data sets.
+Code review processes are also a great way to share knowledge across the team. The reviewer will oftentimes share
+valuable insights on the code that you might not have thought of.
+
+How thorough the review should be depends on the amount of automated testing. If you have a comprehensive test suite
+that everyone on the team has confidence in, then it is not necessary to have a thorough review process. On the other
```

Comment:
> These are hard things to test for that a thorough and well executed code review will highlight.

I was trying to get at the fact that infrastructure code requires an even more thorough review to look for and account for bugs because you don't have automated testing, but you are right that I probably went too overboard in the claim here.

Removed as Jim suggested to avoid the politics: 9e42e51